src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -95,28 +85,121 @@ public class PublishPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
+                    propertiesDictionary.Add(FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath);
+
+                    var artifact = await buildHelper.AssociateArtifact(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
+                    var parallel = context.GetInput(ArtifactEventProperties.Parallel, required: false);
+
+                    var parallelCount = 1;
+                    if(parallel == "true") 
+                    {
+                        parallelCount = GetParallelCount(context, context.GetInput(ArtifactEventProperties.ParallelCount, required: false));
+                    }
 
-            // Upload to VSTS BlobStore, and associate the artifact with the build.
-            context.Output(StringUtil.Loc("UploadingPipelineArtifact", fullPath, buildId));
-            PipelineArtifactServer server = new PipelineArtifactServer();
-            await server.UploadAsync(context, projectId, buildId, artifactName, fullPath, token);
-            context.Output(StringUtil.Loc("UploadArtifactFinished"));
+                    // To copy all the files in one directory to another directory.
+                    // Get the files in the source folder. (To recursively iterate through
+                    // all subfolders under the current directory, see
+                    // "How to: Iterate Through a Directory Tree.")
+                    // Note: Check for target path was performed previously
+                    // in this code example.
+                    if (System.IO.Directory.Exists(fileSharePath))
+                    {
+                        await FileShareProvider.DirectoryCopyWithMiniMatch(targetPath, artifactPath, context, parallelCount);
+                        context.Output(StringUtil.Loc("CopyFileComplete", artifactPath));
+                    }
+                   
+                }
+                else 
+                {
+                    // file share artifacts are not currently supported on OSX/Linux.
+                    throw new InvalidOperationException(StringUtil.Loc("FileShareOperatingSystemNotSupported"));
+                }
+            }
         }
-     
+
         private string NormalizeJobIdentifier(string jobIdentifier)
         {
             jobIdentifier = jobIdentifierRgx.Replace(jobIdentifier, string.Empty).Replace(".default", string.Empty);
             return jobIdentifier;
         }
+
+        private int GetParallelCount(AgentTaskPluginExecutionContext context, string parallelCount)
+        {
+            var result = 8;
+            if(int.TryParse(parallelCount, out result))
+            {
+                if(result < 1) {
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
+                    result = 1;
+                }else if(result > 128){
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
+                    result = 128;
```

Comment:
where did these numbers come from?

Reply:
The is how build artifact task is doing.